### ISSUE TITLE
Vllm's sampling significantly more deterministic than HF?

### ISSUE THREAD

**Issue by syskn** [vllm-project/vllm#712]
### Moving from Discussions https://github.com/vllm-project/vllm/discussions/471

Hi guys, great work!

I have been experimenting with the library for several weeks, and immediately noticed that sampled tokens (with the same temperature and such) are significantly more deterministic with Vllm vs. HF Transformers using the same models - with temperature lower than 0.7, often the first 5-10 sampled tokens are exactly same across few different generations, even recreating the original text in the datasets verbatim, like there is some greedy decoding going on (when it is not). This unfortunately leads to a significant repetition issue I've never seen with HF.

The issue is not related to special tokens such as <s></s>. I modified vllm so that it never generates those special tokens like HF's bad_words_ids, but the issue persists (those special tokens will also make the inference quality significantly worse especially with non-chat prompts, but it is a different issue).

In the mean time, I have also been checking and modifying codebase to see if there is any discrepancy between the sampling process, but I am not sure about the difference.

Does anyone else experienced a similar behavior?

https://github.com/vllm-project/vllm/issues/590: Related topic with actual examples. The topic is about GPTJ, but the issue is there with other architectures such as Llama and NeoX.

**Comment by BaiMoHan** [comment]
Yes.If my memory serves right, this problem arose when I was using version 0.1.1.

**Comment by luohao123** [comment]
I got exactly same issue here!


**the speed are faster, but the performance dropped a lot**

the final output are really not like hf even genering params are same!!

**Comment by ehartford** [comment]
I had the same experience.  I set up a client with VLLM and they saw totally different (and very disappointing) output from VLLM vs the original huggingface code, using the exact same parameters - temperature, repetition penalty, etc.

**Comment by Sanster** [comment]
> I had the same experience. I set up a client with VLLM and they saw totally different (and very disappointing) output from VLLM vs the original huggingface code, using the exact same parameters - temperature, repetition penalty, etc.

Hi, I noticed that the default value of `repetition_penalty` in HF is 1.0, but in vllm the default value of `presence_penalty` is 0. So should we set `presence_penalty = repetition_penalty - 1.0`? Thanks

**Comment by syskn** [comment]
> > I had the same experience. I set up a client with VLLM and they saw totally different (and very disappointing) output from VLLM vs the original huggingface code, using the exact same parameters - temperature, repetition penalty, etc.
> 
> Hi, I noticed that the default value of `repetition_penalty` in HF is 1.0, but in vllm the default value of `presence_penalty` is 0. So should we set `presence_penalty = repetition_penalty - 1.0`? Thanks

This should not be the case here. HF-style repetition penalty works like this:
value == 1.0 bypass
value > 1.0 penaltize repetition
value < 1.0 "promote" repetition

Whereas OpenAI-style frequency penalty and presence penalty are:
value == 0 bypass
value > 0 penaltize repetition

**Comment by Sanster** [comment]
> promote

Yes, I also encountered the issue of sampling. In my testing, I used HF-style repetition_penalty. However, even after changing presence_penalty to repetition_penalty-1, the problem still persists.

**Comment by Sanster** [comment]
> > > I had the same experience. I set up a client with VLLM and they saw totally different (and very disappointing) output from VLLM vs the original huggingface code, using the exact same parameters - temperature, repetition penalty, etc.
> > 
> > 
> > Hi, I noticed that the default value of `repetition_penalty` in HF is 1.0, but in vllm the default value of `presence_penalty` is 0. So should we set `presence_penalty = repetition_penalty - 1.0`? Thanks
> 
> This should not be the case here. HF-style repetition penalty works like this (it's a bit strange implementation for historical reason): value == 1.0 bypass value > 1.0 penaltize repetition value < 1.0 "promote" repetition
> 
> Whereas OpenAI-style frequency penalty and presence penalty are: value == 0 bypass value > 0 penaltize repetition value < 0 "promote" repetition

Yes, I also encountered the issue of sampling. In my testing, I used HF-style repetition_penalty. However, even after changing presence_penalty to repetition_penalty-1, the problem still persists.

**Comment by syskn** [comment]
Actually, there is also a difference between how each repetition penalties are implemented (HF is exponential, OpenAI/vLLM is linear, presence penalty is one-off). However, this is not directly related to the potential degradation issue from what I understand.

**Comment by Linzecong** [comment]
> I got exactly same issue here!
> 
> **the speed are faster, but the performance dropped a lot**
> 
> the final output are really not like hf even genering params are same!!

Subjectively feel that the HF generation will be better. I'm trying to adjust various parameters to align at low temperature.

**Comment by luohao123** [comment]
On my side, I didn't care about which params, I just want aligned with hf, since hf result are good, but vllm worse..

**Comment by desperadoola** [comment]
https://github.com/vllm-project/vllm/blob/1b151ed181f5ef0d267c1cf6d15d6f6f43e302ed/vllm/model_executor/models/gpt_neox.py#L70-L72

I'm wondering whether `PagedAttentionWithRoPE` is initialized using correct model max positions. Since the default max positions for `PagedAttentionWithRoPE` is 8192. 

https://github.com/vllm-project/vllm/blob/1b151ed181f5ef0d267c1cf6d15d6f6f43e302ed/vllm/model_executor/layers/attention.py#L246-L280

**Comment by kechunFIVE** [comment]
> On my side, I didn't care about which params, I just want aligned with hf, since hf result are good, but vllm worse..

vllm also seems to be getting worse in my experiments, but your avatar is so cool.

**Comment by Sanster** [comment]
Share some of my experimental results.

## LongEval result
Here is the result of my running the [longeval](https://github.com/DachengLi1/LongChat) `Fine-grained Line Retrieval task`, using same model and same parameters.

- temperature: 0.01
- top_p: 0.95

<img width="849" alt="image" src="https://github.com/vllm-project/vllm/assets/3998421/0e348486-0cab-4c1a-8382-c350dacbec72">

## Ceval
However, the same model, tested on the ceval task using HF and VLLM, did not show a significant difference.

**Comment by syskn** [comment]
Sanster,

Wow. Just to be sure, did you merge RoPE interpolation support PR (either https://github.com/vllm-project/vllm/pull/555 or  https://github.com/vllm-project/vllm/pull/690 respectively) which is not yet added to the latest release?

AFAIK, if a rope-encoded model is 2048 ctx length, eval starts to significantly degrade from around 2400 tokens without interpolation methods.

**Comment by creatorrr** [comment]
We are also seeing the same effect for longer sequences. What could be the issue @zhuohan123 ?

**Comment by Sanster** [comment]
> Sanster,
> 
> Wow. Just to be sure, did you merge RoPE interpolation support PR (either #555 or #690 respectively) which is not yet added to the latest release?
> 
> AFAIK, if a rope-encoded model is 2048 ctx length, eval starts to significantly degrade from around 2400 tokens without interpolation methods.

My model is baichuan-13b (Alibi location embedding) sft, trained with a 4096 ctx length, so the blue line (HF forward) is normal to me, however, the orange line (VLLM forward) is quite confusing for me.

**Comment by luohao123** [comment]
I think vllm doesn't support alibi context expanding?

However, since you are trained with 4096, so the drop is really mysterious.

**Comment by heshuguo** [comment]
yes, the same issue. Get bad result after using vllm. The model is baichuan-13b. not sure how to fix the issue.

**Comment by Sanster** [comment]
> Share some of my experimental results.
> 
> ## LongEval result
> Here is the result of my running the [longeval](https://github.com/DachengLi1/LongChat) `Fine-grained Line Retrieval task`, using same model and same parameters.
> 
> * temperature: 0.01
> * top_p: 0.95
> * vllm max_num_batched_tokens: 4096
> 
> <img alt="image" width="849" src="https://user-images.githubusercontent.com/3998421/262851274-0e348486-0cab-4c1a-8382-c350dacbec72.png">
> ## Ceval
> However, the same model, tested on the ceval task using HF and VLLM, did not show a significant difference.

I finally found out how to fix this problem. For baichuan-13b(or other model using ALiBi), you can try change following line in `vllm/model_executor/layers/attention.py` 

https://github.com/vllm-project/vllm/blob/320a622ec4d098f2da5d097930f4031517e7327b/vllm/model_executor/layers/attention.py#L354

change to HF implement:

```python
bias = bias[None, :].repeat(prompt_len, 1)
```

**Comment by luohao123** [comment]
@Sanster Can u make an PR?

**Comment by Sanster** [comment]
> @Sanster Can u make an PR?

I'm not sure how this modification will affect other models or datasets. If more people could share the impact of this modification, it might be worth proposing a PR.

**Comment by Sanster** [comment]
This commit https://github.com/vllm-project/vllm/commit/db09d4ad833b1d8911f14852a484790c53818a5d seems to have fixed the problem with alibi. In my tests, the results are very close to just modifying `bias = bias[None, :].repeat(prompt_len, 1)`.

**Comment by diogobragaswogo** [comment]
I can confirm this quality degradation is also present in Falcon-7B. A finetuned model whose BLEU score is ~94 degrades to a BLEU score of ~80.

In my case, this is a blocker for using vLLM, because:

1.  The quality of the model must be high.
2. Given it is a difference in vLLM's algorithm and not the model, it makes it virtually impracticable and/or impossible to improve the model's performance during training.

For the mean time, I'll be forced to use TGI but I'd like to use vLLM if this issue is fixed.

**Comment by yanxiyue** [comment]
after merge pr #866 (
Add repetition_penalty aligned with huggingface) to this fork https://github.com/yanxiyue/vllm, vllm works well for me on baichuan13b

**Comment by yungangwu** [comment]
I have the same problem on baichuan2. Very serious inference performance degradation in baichuan2 with use vllm.  I'm currently using version is 0.2.1. Has anyone solved this problem?

**Comment by hmellor** [comment]
Closing this issue as stale as there has been no discussion in the past 3 months.

If you are still experiencing the issue you describe, feel free to re-open this issue.

**Comment by hmellor** [comment]
Based on the last discussion before the issue when stale, `repetition_penalty` was the cause of the discrepancy. 

This feature was added in https://github.com/vllm-project/vllm/pull/1424.
